---
title: "CloneID Use Case 3"
author: "Thomas Veith, Noemi Andor"
date: "4/11/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE)
knitr::opts_knit$set(root.dir = "/Users/4470246/Projects/SoftwarePackages_InHouse/CLONEID/code")
```

## USE CASE 3 (In-vitro genotype-phenotype association studies)

Here we are gathering various morphological and genotypic features from various lineages culture in-vitro in our lab.

The input to this section are the cell line lineages.

The output are two table: one containing the copy number per genomic region for each clone in the lineage, and another containing various morphological features inferred from imaging and cell segmentation.

```{r}
## USE CASE 3 (In-vitro genotype-phenotype association studies)
library(cloneid)
lineageCLs <-c("NUGC-4_A4_seed","NCI-N87_A59_seed","HGC-27_A22_seed","KATOIII_A6_seed","SNU-668_A9_seed","MKN-45_A4_seed","SNU-638_A4_seed","SNU-601_A4_seed");
cl = lineageCLs[5]
out=findAllDescendandsOf(ids=cl, recursive = F)
ii=3
```

```{r}
## GenomePerspective
mydb = cloneid::connect2DB()
## Do we have Genome sequencing data for any lineage from this cell line?
stmt = paste0("select distinct origin from Perspective where whichPerspective='GenomePerspective' and sampleSource = '",unique(out$cellLine),"'")
rs = suppressWarnings(dbSendQuery(mydb, stmt))
origin=fetch(rs, n=-1)[,"origin"]

## Download genomic profile for one subpopulation:
sps=getSubclones(cloneID_or_sampleName = origin, whichP = "GenomePerspective")
p=sapply(names(sps), function(x) getSubProfiles(cloneID_or_sampleName = as.numeric(extractID(x)), whichP = "GenomePerspective"))
clonemembership=unlist(sapply(names(p), function(x) rep(x, ncol(p[[x]]))))
clonesizes = sapply(p,ncol)
pie(clonesizes)
p=do.call(cbind, p)
print(dim(p))
print(head(p[,1:min(3,ncol(p))] ))

## Mimic bulk data:
# p=p[,1:5]

## UMAP visualization
if(ncol(p)>20){
  col=rainbow(length(unique(clonemembership)))
  names(col)=unique(clonemembership)
  la=umap::umap(t(p))
  plot(la$layout,pch=20, col=col[clonemembership])
}else{
  gplots::heatmap.2(t(p))
  ## @TODO: Not sure if this still works if p is a vector rather than a matrix -- would need to check to make sure there is no error
}
```


```{r}
## TranscriptomePerspective: same as for GenomePerspective
```



```{r}
## MorphologyPerspective
sps=getSubclones(cloneID_or_sampleName = out$id[ii],whichP = "MorphologyPerspective")
sps=extractID(names(sps))
p=sapply(sps, function(x) getSubProfiles(cloneID_or_sampleName = as.numeric(x), whichP = "MorphologyPerspective"), simplify = F)
p=do.call(cbind, p)
print(dim(p))
print(head(p[,1:min(3,ncol(p))] ))

## UMAP visualization
la=umap::umap(t(p))
plot(la$layout,pch=20)
```



```{r}
## GenomePerspective Bulk
library(cloneid)
out=findAllDescendandsOf("P06270", recursive = T)
mydb = cloneid::connect2DB()
stmt = paste0("select distinct origin from Perspective where whichPerspective='GenomePerspective' and origin IN ('",paste(unique(out$id), collapse="','"),"')")
rs = suppressWarnings(dbSendQuery(mydb, stmt))
origin=fetch(rs, n=-1)[,"origin"]

p=sapply(origin, function(x) getSubProfiles(cloneID_or_sampleName = x, whichP = "GenomePerspective"))
p=do.call(cbind, p)
gplots::heatmap.2(t(p), trace = "n")
```



```{r}
## Timeline for internal nodes (seedings)
la=findAllDescendandsOf(id="SNU-668_A1_seed");
la=la[la$event=="seeding",]
la=sapply(la$id, function(x) findAllDescendandsOf(id=x, recursive=F), simplify=F)
la=sapply(la, function(x) quantile(as.POSIXct(x$date), c(0,1)),simplify=F)
## This comes closer to what we want:
barplot(sapply(la, function(x) x[2]-x[1]))
## ... but is not actually there yet for two reasons:
# 1) x[2]-x[1] returns the time difference in different units (hours or days depending on the magnitude of the difference). We would want everything to be the same unit of course. 
# 2) the bars are equally spaced, unfortunately "barplot" doesn't have a parameter "at", specifying the x-axis locations. We would need a barplot function that allows us to specify where the bars should be at along the x-axis (e.g. at = sapply(la,"[",1, simplify=F) ).
```



```{r}
## Record "genotypic" information into CLONEID
## Note: this includes cell morphology info. Referring to a morphology feature as “genotype” is technically not correct. We will need to come up with a better name
library(cloneid)
library(matlab)
data(CloneProfiles)
for(p in names(CloneProfiles)){
  OUT=paste0("~/Downloads/testViewPerspective",filesep,p)
  dir.create(OUT,recursive = T)
  for(name in names(CloneProfiles[[p]])){
    write.table(CloneProfiles[[p]][[name]],file=paste0(OUT,filesep,name),sep="\t",quote=F,row.names = F)
  }
  name=grep("spstats",names(CloneProfiles[[p]]),value=T)
  suffix=grep("spstats",names(CloneProfiles[[p]]),invert = T, value=T)[1]
  suffix=gsub(fileparts(name)$name,"",suffix)
  viewPerspective(spstatsFile =paste0(OUT,filesep,name), whichP = p,suffix = suffix)
}

```


```{r}
## Record phenotypic information into CLONEID (image segmentation based cell counts)
dir.create("~/CellSegmentations")
id="HGC-27_A1_seedTP48"
f=list.files("path to code_examples4webportal", pattern=id, full.names = T)
sapply(f, function(x) file.copy(x, paste0("~/CellSegmentations/",matlab::fileparts(x)$name)))
## This will segment the images, and ask user if segmentation looks ok. User can check this by looking at the masks saved under "~/Downloads/tmp/vis/". CLONEID will use the approved images to extrapolate cell counts to entire flask and save result into the database: 
harvest(id,from="HGC-27_A1_seed", preprocessing = F)
# Note that result will not be saved because the entry (id) already exists in the DB. 
```